File: source/extensions/Physics/PhysicsDebugger.java
package source.extensions.Physics;

import com.badlogic.gdx.graphics.OrthographicCamera;
import com.badlogic.gdx.physics.box2d.Box2DDebugRenderer;
import com.badlogic.gdx.utils.viewport.ExtendViewport;
import com.badlogic.gdx.utils.viewport.Viewport;

/**
 * @author Xico
 */

public class PhysicsDebugger {

    private final Box2DDebugRenderer renderer;

    /**
     * As long as {@link PhysicsDebugger#resize(int, int)} gets called, the b2dCamera + b2dViewport will
     * keep the display working as intended, even after resize events.
     */
    private final OrthographicCamera b2dCamera;
    private final Viewport b2dViewport;

    public PhysicsDebugger(int worldWidth, int worldHeight) {
        renderer = new Box2DDebugRenderer();
        final float vpWidth = Utils.toB2DUnits(worldWidth);
        final float vpHeight = Utils.toB2DUnits(worldHeight);
        b2dCamera = new OrthographicCamera();
        b2dCamera.setToOrtho(false, vpWidth, vpHeight);
        b2dViewport = new ExtendViewport(vpWidth, vpHeight, b2dCamera);
    }

    /**
     * This is crucial to update the [b2dViewport] when a resize event occurs.
     *
     * @param width new width.
     * @param height new height.
     */
    void resize(int width, int height) {
        b2dViewport.update(width, height);
    }

    void render(PhysicsWorld physicsWorld) {
        renderer.render(physicsWorld.getWorld(), b2dCamera.combined);
    }
}

File: kot/physics/PhysicsDebugger.kt
package kot.physics

import com.badlogic.gdx.graphics.OrthographicCamera
import com.badlogic.gdx.physics.box2d.Box2DDebugRenderer
import com.badlogic.gdx.utils.viewport.ExtendViewport
import com.badlogic.gdx.utils.viewport.Viewport

internal class PhysicsDebugger(worldWidth: Int, worldHeight: Int) {

  // TODO: Accept the injection of different viewports.
  // TODO: They will need to have their dimensions converted to B2D units, though.

  /**
   * The [Box2DDebugRenderer] instance that will allow the rendering
   * of the given [PhysicsWorld].
   */
  private val renderer: Box2DDebugRenderer

  /**
   * Dedicated [OrthographicCamera] for the Box2D rendering, because of the
   * scaling issues. Will be managed by a dedicated [Viewport] as well.
   */
  private val camera: OrthographicCamera

  /**
   * Dedicated [Viewport] that will manage the [camera] instance.
   */
  private val viewport: Viewport

  init {
    renderer = Box2DDebugRenderer()
    val viewportWidth: Float = (worldWidth.toFloat()).asBox2DUnits()
    val viewportHeight: Float = (worldHeight.toFloat()).asBox2DUnits()
    camera = OrthographicCamera().apply {
      setToOrtho(false, viewportWidth, viewportHeight)
    }
    viewport = ExtendViewport(viewportWidth, viewportHeight, camera)
  }

  /**
   * Called on [PhysicsWorld.resize] to make sure that the [viewport]
   * responds correctly to the screen resizing events.
   *
   * @param width   new screen width.
   * @param height  new screen height.
   */
  internal fun resize(width: Int, height: Int) = viewport.update(width, height)

  /**
   * Calls [Box2DDebugRenderer.render] on the given [PhysicsWorld.world], using the
   * dedicated [camera]'s projection and view matrix.
   *
   * @param world instance of the [PhysicsWorld] that will be rendered.
   */
  internal fun render(world: PhysicsWorld) = renderer.render(world.world, camera.combined)
}
